src/app/clusters/time-synchronization-server/TimeSyncDataProvider.cpp

```diff
@@ -0,0 +1,226 @@
+
+    TLV::TLVReader reader;
+
+    reader.Init(bufferSpan.data(), bufferSpan.size());
+    ReturnErrorOnFailure(reader.Next(TLV::AnonymousTag()));
+    ReturnErrorOnFailure(timeSource.Decode(reader));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR TimeSyncDataProvider::ClearTrustedTimeSource()
+{
+    return mPersistentStorage->SyncDeleteKeyValue(DefaultStorageKeyAllocator::TSTrustedTimeSource().KeyName());
+}
+
+CHIP_ERROR TimeSyncDataProvider::StoreDefaultNtp(const CharSpan & defaultNtp)
+{
+    return mPersistentStorage->SyncSetKeyValue(DefaultStorageKeyAllocator::TSDefaultNTP().KeyName(), defaultNtp.data(),
+                                               static_cast<uint16_t>(defaultNtp.size()));
+}
+
+CHIP_ERROR TimeSyncDataProvider::LoadDefaultNtp(MutableByteSpan & defaultNtp)
+{
+    return Load(DefaultStorageKeyAllocator::TSDefaultNTP().KeyName(), defaultNtp);
+}
+
+CHIP_ERROR TimeSyncDataProvider::ClearDefaultNtp()
+{
+    return mPersistentStorage->SyncDeleteKeyValue(DefaultStorageKeyAllocator::TSDefaultNTP().KeyName());
+}
+
+CHIP_ERROR TimeSyncDataProvider::StoreTimeZone(const TimeZone & timeZoneList)
+{
+    uint8_t buffer[kTimeZoneListMaxSerializedSize];
+    TLV::TLVWriter writer;
+    TLV::TLVType outerType;
+
+    writer.Init(buffer);
+    ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Array, outerType));
+
+    for (auto const & timeZone : timeZoneList)
+    {
+        ReturnErrorOnFailure(timeZone.Encode(writer, TLV::AnonymousTag()));
+    }
+
+    ReturnErrorOnFailure(writer.EndContainer(outerType));
+
+    return mPersistentStorage->SyncSetKeyValue(DefaultStorageKeyAllocator::TSTimeZone().KeyName(), buffer,
+                                               static_cast<uint16_t>(writer.GetLengthWritten()));
+}
+CHIP_ERROR TimeSyncDataProvider::LoadTimeZone(TimeZone & timeZoneList, uint8_t & size)
+{
+    uint8_t buffer[kTimeZoneListMaxSerializedSize];
+    MutableByteSpan bufferSpan(buffer);
+    size           = 0;
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    ReturnErrorOnFailure(Load(DefaultStorageKeyAllocator::TSTimeZone().KeyName(), bufferSpan));
+
+    TLV::TLVReader reader;
+    TLV::TLVType outerType;
+
+    reader.Init(bufferSpan.data(), bufferSpan.size());
+    ReturnErrorOnFailure(reader.Next(TLV::TLVType::kTLVType_Array, TLV::AnonymousTag()));
+    ReturnErrorOnFailure(reader.EnterContainer(outerType));
+    auto tz   = timeZoneList.begin();
+    uint8_t i = 0;
+
+    while (reader.Next() != CHIP_ERROR_END_OF_TLV && i < timeZoneList.size())
+    {
+        app::Clusters::TimeSynchronization::Structs::TimeZoneStruct::Type timeZone;
+        ReturnErrorOnFailure(timeZone.Decode(reader));
+        tz[i].offset  = timeZone.offset;
+        tz[i].validAt = timeZone.validAt;
+        if (timeZone.name.HasValue())
+        {
+            if (timeZone.name.Value().size() <= tz[i].name.Value().size())
```

Comment:
In this specific case because we always use it initialised. I am not anticipating a situation where this code could be used by another time sync implementation. But I will make it safer by adding a check for it.

Reply:
See my other comments below: it's not in fact always initialized as things stand; we just don't have good enough error case test coverage.

Reply:
See my other comments below: it's not in fact always initialized as things stand; we just don't have good enough error case test coverage.